### PR TITLE
Use automatic memory management with GSL objects.

### PR DESCRIPTION
**Pull request by bangerth**
I don't have GSL installed, so don't know whether that actually works. Do any of our testers check this?

/rebuild

### CODE REVIEW

**Comment by masterleinad** [comment]
```
/home/runner/work/dealii/dealii/source/base/function_cspline.cc: In constructor ‘dealii::Functions::CSpline<dim>::CSpline(const std::vector<double>&, const std::vector<double>&)’:
/home/runner/work/dealii/dealii/source/base/function_cspline.cc:58:22: error: cannot convert ‘std::unique_ptr<gsl_spline, void (*)(gsl_spline*)>*’ to ‘gsl_spline*’ for argument ‘1’ to ‘int gsl_spline_init(gsl_spline*, const double*, const double*, size_t)’
                     n);
                      ^
/home/runner/work/dealii/dealii/source/base/function_cspline.cc: In member function ‘virtual double dealii::Functions::CSpline<dim>::value(const dealii::Point<dim>&, unsigned int) const’:
/home/runner/work/dealii/dealii/source/base/function_cspline.cc:79:43: error: cannot convert ‘std::unique_ptr<gsl_spline, void (*)(gsl_spline*)>* const’ to ‘const gsl_spline*’ for argument ‘1’ to ‘double gsl_spline_eval(const gsl_spline*, double, gsl_interp_accel*)’
     return gsl_spline_eval(cspline, x, acc);
                                           ^
/home/runner/work/dealii/dealii/source/base/function_cspline.cc: In member function ‘virtual dealii::Tensor<1, dim> dealii::Functions::CSpline<dim>::gradient(const dealii::Point<dim>&, unsigned int) const’:
/home/runner/work/dealii/dealii/source/base/function_cspline.cc:100:65: error: cannot convert ‘std::unique_ptr<gsl_spline, void (*)(gsl_spline*)>* const’ to ‘const gsl_spline*’ for argument ‘1’ to ‘double gsl_spline_eval_deriv(const gsl_spline*, double, gsl_interp_accel*)’
     const double   deriv = gsl_spline_eval_deriv(cspline, x, acc);
                                                                 ^
/home/runner/work/dealii/dealii/source/base/function_cspline.cc: In member function ‘virtual double dealii::Functions::CSpline<dim>::laplacian(const dealii::Point<dim>&, unsigned int) const’:
/home/runner/work/dealii/dealii/source/base/function_cspline.cc:124:50: error: cannot convert ‘std::unique_ptr<gsl_spline, void (*)(gsl_spline*)>* const’ to ‘const gsl_spline*’ for argument ‘1’ to ‘double gsl_spline_eval_deriv2(const gsl_spline*, double, gsl_interp_accel*)’
     return gsl_spline_eval_deriv2(cspline, x, acc);
                                                  ^
/home/runner/work/dealii/dealii/source/base/function_cspline.cc: In instantiation of ‘dealii::Functions::CSpline<dim>::CSpline(const std::vector<double>&, const std::vector<double>&) [with int dim = 1]’:
/home/runner/work/dealii/dealii/source/base/function_cspline.cc:151:18:   required from here
/home/runner/work/dealii/dealii/source/base/function_cspline.cc:30:30: error: no matching function for call to ‘std::unique_ptr<gsl_interp_accel, void (*)(gsl_interp_accel*)>::unique_ptr()’
     , interpolation_values(y_)
                              ^
```

**Review by masterleinad — APPROVED** [review]
Looks good to me.

**Comment by masterleinad** [comment]
```
/jenkins/workspace/dealii-tidy_PR-12324/source/base/function_cspline.cc:31:7: error: no matching constructor for initialization of 'std::unique_ptr<gsl_interp_accel, void (*)(gsl_interp_accel *)>' [clang-diagnostic-error]
/jenkins/workspace/dealii-tidy_PR-12324/source/base/function_cspline.cc:32:7: error: no matching constructor for initialization of 'std::unique_ptr<gsl_spline, void (*)(gsl_spline *)>' [clang-diagnostic-error]
```

File: source/base/function_cspline.cc [hunk 1]
```diff
@@ -28,6 +28,8 @@ namespace Functions
                         const std::vector<double> &y_)
     : interpolation_points(x_)
     , interpolation_values(y_)
+    , acc(nullptr)
+    , cspline(nullptr)
```
Comment (masterleinad) [inline source/base/function_cspline.cc:32]:
```suggestion
    , acc(gsl_interp_accel_alloc(), 
          [](gsl_interp_accel *p) { gsl_interp_accel_free(p); })
    , cspline(gsl_spline_alloc(gsl_interp_cspline, interpolation_points.size()), 
              [](gsl_spline *p) { gsl_spline_free(p); })
```

File: source/base/function_cspline.cc [hunk 2]
```diff
@@ -43,29 +45,23 @@ namespace Functions
                                   interpolation_points[i],
                                   interpolation_points[i + 1]));
 
-    acc                  = gsl_interp_accel_alloc();
+    acc = decltype(acc)(gsl_interp_accel_alloc(),
+                        [](gsl_interp_accel *p) { gsl_interp_accel_free(p); });
+
```
Comment (masterleinad) [inline source/base/function_cspline.cc:50]:
```suggestion
```

File: source/base/function_cspline.cc [hunk 3]
```diff
@@ -43,29 +45,23 @@ namespace Functions
...
     const unsigned int n = interpolation_points.size();
-    cspline              = gsl_spline_alloc(gsl_interp_cspline, n);
+
+    cspline = decltype(cspline)(gsl_spline_alloc(gsl_interp_cspline, n),
+                                [](gsl_spline *p) { gsl_spline_free(p); });
+
```
Comment (masterleinad) [inline source/base/function_cspline.cc:55]:
```suggestion
```

**Comment by bangerth** [comment]
Yes, nice and elegant! Thanks for the suggestion, @masterleinad !

**Pull request merged**